src/Proxy/DriveItemProxy.php

```diff
@@ -255,45 +257,86 @@ public function delete()
      * Uploads a file under this folder drive item.
      *
      * @param string $name
-     *        The name.
-     * @param string|resource|\GuzzleHttp\Psr7\Stream $content
-     *        The content.
+     *        The name of file(relative file path if in same directory (or) absolute file path).
      * @param array $options
      *        The options.
      *
      * @return DriveItemProxy
      *         The drive item created.
      *
-     * @todo Support name conflict behavior.
-     * @todo Support content type in options.
      */
-    public function upload($name, $content, array $options = [])
+    public function upload($name, array $options = [])
     {
         $name         = rawurlencode($name);
         $driveLocator = "/drives/{$this->parentReference->driveId}";
         $itemLocator  = "/items/{$this->id}";
-        $endpoint     = "$driveLocator$itemLocator:/$name:/content";
-
-        $body = $content instanceof Stream ?
-            $content
-            : Psr7\stream_for($content);
-
-        $response = $this
+        $uploadSessionEndpoint = "$driveLocator$itemLocator:/$name:/createUploadSession";
+        //include extended options with existing item for upload session
+        $item = array_merge(
+            array("name" => $name, 'description' => '' ,"@microsoft.graph.conflictBehavior" => "rename"), $options
+        );
+
+		//create an upload session for large files
+		$uploadSessionResponse = $this
             ->graph
-            ->createRequest('PUT', $endpoint)
-            ->addHeaders($options)
-            ->attachBody($body)
+            ->createRequest('POST', $uploadSessionEndpoint)
+            ->addHeaders(array('Content-Type' => 'application/json'))
+            ->attachBody(array("item" =>$item))
             ->execute();
-
-        $status = $response->getStatus();
-
-        if ($status != 201) {
-            throw new \Exception("Unexpected status code produced by 'PUT $endpoint': $status");
-        }
-
-        $driveItem = $response->getResponseAsObject(DriveItem::class);
-
-        return new self($this->graph, $driveItem);
+		
+		$uploadSessionResponseBody=$uploadSessionResponse->getBody();
+
+        if(is_array($uploadSessionResponseBody) && array_key_exists("uploadUrl", $uploadSessionResponseBody)) {
+
+            $uploadUrl = $uploadSessionResponseBody['uploadUrl'];
+
+			$handle = @fopen($name, "r");
+			if ($handle) {
+				$bytesRead=0;
+				$buffer = "";
+				$totalSize=filesize($name);
```

Comment:
This would also be tricky when you will work with a stream, I would suggest that this be also handed to your function by the caller (via `$options`). If it's a file, they'll do what you did, if it's a string, they'll do a `strlen` for you, and so on.

Reply:
added total size configurable via options. If it doesn't find one such option, it'll calculate the length from stream